R/lazyframe-frame.R

```diff
@@ -358,12 +755,1924 @@ lazyframe__slice <- function(offset, length = NULL) {
+    closed = "left",
+    label = "left",
+    group_by = NULL,
+    start_by = "window") {
+  wrap({
+    check_dots_empty0(...)
+    closed <- arg_match0(closed, values = c("both", "left", "right", "none"))
+    start_by <- arg_match0(
+      start_by,
+      values = c(
+        "window", "datapoint", "monday", "tuesday", "wednesday", "thursday",
+        "friday", "saturday", "sunday"
+      )
+    )
+    every <- parse_as_duration_string(every)
+    offset <- parse_as_duration_string(offset) %||% "0ns"
+    period <- parse_as_duration_string(period) %||% every
+    group_by <- parse_into_list_of_expressions(!!!group_by)
+
+    self$`_ldf`$group_by_dynamic(
+      as_polars_expr(index_column)$`_rexpr`, every, period, offset, label,
+      include_boundaries, closed,
+      group_by, start_by
+    )
+  })
+}
+
+#' Plot the query plan
+#'
+#' This only returns the "dot" output that can be passed to other packages, such
+#' as `DiagrammeR::grViz()`.
+#'
+#' @param ... Not used..
+#' @param optimized Optimize the query plan.
+#' @inheritParams lazyframe__explain
+#'
+#' @return A character vector
+#'
+#' @examples
+#' lf <- pl$LazyFrame(
+#'   a = c("a", "b", "a", "b", "b", "c"),
+#'   b = 1:6,
+#'   c = 6:1
+#' )
+#'
+#' query <- lf$group_by("a", maintain_order = TRUE)$agg(
+#'   pl$all()$sum()
+#' )$sort(
+#'   "a"
+#' )
+#'
+#' query$to_dot() |> cat()
+#'
+#' # You could print the graph by using DiagrammeR for example, with
+#' # query$to_dot() |> DiagrammeR::grViz().
+lazyframe__to_dot <- function(
+    ...,
+    optimized = TRUE,
+    type_coercion = TRUE,
+    predicate_pushdown = TRUE,
+    projection_pushdown = TRUE,
+    simplify_expression = TRUE,
+    slice_pushdown = TRUE,
+    comm_subplan_elim = TRUE,
+    comm_subexpr_elim = TRUE,
+    cluster_with_columns = TRUE,
+    streaming = FALSE) {
+  lf <- self |>
+    self$`_ldf`$optimization_toggle(
+      pe_coercion = type_coercion,
+      predicate_pushdown = predicate_pushdown,
+      projection_pushdown = projection_pushdown,
+      simplify_expression = simplify_expression,
+      slice_pushdown = slice_pushdown,
+      comm_subplan_elim = comm_subplan_elim,
+      comm_subexpr_elim = comm_subexpr_elim,
+      cluster_with_columns = cluster_with_columns,
+      streaming = streaming,
+      eager = FALSE
+    )
+
+  self$`_ldf`$to_dot(optimized)
+}
+
+#' Create an empty or n-row null-filled copy of the LazyFrame
+#'
+#' Returns a n-row null-filled LazyFrame with an identical schema. `n` can be
+#' greater than the current number of rows in the LazyFrame.
+#'
+#' @param n Number of (empty) rows to return in the cleared frame.
+#'
+#' @return A n-row null-filled LazyFrame with an identical schema
+#'
+#' @examples
+#' df <- pl$LazyFrame(
+#'   a = c(NA, 2, 3, 4),
+#'   b = c(0.5, NA, 2.5, 13),
+#'   c = c(TRUE, TRUE, FALSE, NA)
+#' )
+#'
+#' df$clear()
+#'
+#' df$clear(n = 5)
+lazyframe__clear <- function(n = 0) {
+  pl$DataFrame(schema = self$schema)$clear(n)$lazy()
```

Comment:
Is this correct???

Reply:
It needs a small adjustment since we don't have `schema` but `.schema_overrides` in `pl$DataFrame()` but that's the idea, yes:

https://github.com/pola-rs/polars/blob/91ad299b17c48422131c584f63acedc758581100/py-polars/polars/lazyframe/frame.py#L3238

(Note that this doesn't work because `$clear()` is not implemented for DataFrame yet)

Reply:
[`a1b5f16` (#49)](https://github.com/eitsupi/neo-r-polars/pull/49/commits/a1b5f162f0e2ee3a50fb683f5e1b4a78991f8519)